### PR TITLE
Fix E2E tests on Windows

### PR DESCRIPTION
**Pull request by mokagio**
Update: Rebased on top of #99. The diff is noisy because it includes the changes from #17, too.
Update2: Merged trunk to the branch and solved conflicts. Shouldn't be noisy anymore.
Update3: Merged trunk to the branch and scope of this PR changes to fixing tests for Windows

---

_Works on my machine™_

image

Let's see what CI does...

---

## Proposed Changes

The PR adds E2E tests to Buildkite pipeline.

## Testing Instructions

1. Confirm E2E tests run successfully in Buildkite

## Pre-merge Checklist

- [x] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/components/onboarding.tsx [hunk 1]
```diff
@@ -94,7 +94,10 @@ export default function Onboarding() {
 	);
 
 	return (
-		<div className="flex flex-row flex-grow">
+		<div
+			className="flex flex-row flex-grow"
+			data-testid="onboarding"
+		>
```
Comment (mokagio) [inline src/components/onboarding.tsx]: We could actually remote this, given the current code doesn't use it because I couldn't get it to work.

File: e2e/sites.test.ts [hunk 2]
```diff
@@ -37,19 +82,26 @@ test.describe( 'Servers', () => {
...
 		// Check the site is running
 		const siteContent = new SiteContent( mainWindow, siteName );
+		expect( await siteContent.siteNameHeading ).toHaveText( siteName );
+
+		await siteContent.navigateToTab( 'Settings' );
+
+		expect( await siteContent.frontendButton ).toBeVisible();
 		const frontendUrl = await siteContent.frontendButton.textContent();
-		expect( frontendUrl ).toBeTruthy();
+		expect( frontendUrl ).not.toBeNull();
```
Comment (mokagio) [inline e2e/sites.test.ts]: It might be better to make this "to not be empty"

File: e2e/sites.test.ts [hunk 3]
```diff
@@ -15,20 +16,64 @@ test.describe( 'Servers', () => {
...
-			// If the folder wasn't ever created, a test assertion will have caught this problem
-		}
+
+		[ tmpSiteDir, onboardingTmpSiteDir ].forEach( async ( dir ) => {
+			// Check if path exists first, because tmpSiteDir should have been deleted by the test that deletes the site.
+			if ( await pathExists( dir ) ) {
+				try {
+					await fs.rm( dir, { recursive: true } );
+				} catch {
+					// fail test because we couldn't clean up
```
Comment (mokagio) [inline e2e/sites.test.ts]: Useless comment!
```suggestion
```

File: e2e/page-objects/onboarding.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	private get localPathButton() {
+		return this.locator.getByTestId( 'select-path-button' );
+	}
+
+	// This usually opens an OS folder dialog, except we can't interact with it in playwrite.
+	// In tests the dialog returns the value of the E2E_OPEN_FOLDER_DIALOG environment variable.
+	async clickLocalPathButtonAndSelectFromEnv() {
+		await this.localPathButton.click();
+	}
```
Comment (mokagio) [inline e2e/page-objects/onboarding.ts]: Part of this logic is duplicated from `add-site-modal.ts`. I shall extract it in a `site-form.ts` object, but maybe in a followup PR. Very keen to get E2E back to stable in CI.

File: .buildkite/shared-pipeline-vars [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
Comment (wojtekn) [inline .buildkite/shared-pipeline-vars]: @mokagio it seems we don't use it. Should we remove it? Or did you intend to refactor pipeline to use this file?

**Review by wojtekn — COMMENTED** [review]
@mokagio I merged trunk to the branch, and it's better now as E2E succeeded for Mac now. 

I fixed paths for Windows, but those still fail. @fluiddot any thoughts?

**Comment by wojtekn** [comment]
Also it's worth noting that E2E tests fail for Windows on GH with the same error, so it doesn't seem to be related to Buildkite, but to tests for Windows in general.

File: .buildkite/shared-pipeline-vars (hunk 5 above)
Comment (mokagio) [reply .buildkite/shared-pipeline-vars]: @wojtekn 

This file is `source`'d before calling `buildkite-agent pipeline upload`, and can be used to set up some variables that will be interpolated in the `.yml` pipeline before uploading it.

See https://github.com/Automattic/buildkite-ci/blob/c0d6ade0cc1b7dd5e8cb69beaf1161a28a679156/src/buildkite/pipelines/studio.tf#L22

Maybe I could updated the message below with something like "in CI" or "by CI".

File: .buildkite/shared-pipeline-vars [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+# This file is `source`'d before calling `buildkite-agent pipeline upload`, and can be used
+# to set up some variables that will be interpolated in the `.yml` pipeline before uploading it.
```
Comment (mokagio) [inline .buildkite/shared-pipeline-vars]:
```suggestion
# This file is `source`'d in CI before calling `buildkite-agent pipeline upload`, and can be used
# to set up some variables that will be interpolated in the `.yml` pipeline before uploading it.
```

**Comment by fluiddot** [comment]
> I fixed paths for Windows, but those still fail. @fluiddot any thoughts?

I tried to run the E2E tests locally on a Windows machine and failed due to not being able to launch the Electron. I'll take a look and try to debug the issue.

**UPDATE:** I checked the launch options and saw that the executable path points to `<BINARY_PATH>/Squirrel.exe` instead of `Studio.exe`.

**UPDATE 2:** The Playwright helper that retrieves the app information expects that the Windows build has a single `exe` file ([reference](https://github.com/spaceagetv/electron-playwright-helpers/blob/87b4fbfe971d2d5c620c2ad9864c412d21c2479c/src/find_parse_builds.ts#L269-L272)). In our case, it has two `Squirrel.exe` and `Studio.exe`. This explains why it's trying to run the incorrect executable. I'll push a fix for this.

**Comment by fluiddot** [comment]
Despite [this fix](https://github.com/Automattic/studio/pull/76/commits/0e2ea06d2e1debb098d9af257fb7e192b0c8c89c), the E2E tests keep failing in Windows although they seem specific to the CI environment. At least, I can't reproduce them locally on a Windows machine. My gut feeling is that the app is not being launched, so we'd need to perform the debugging directly there. I'm thinking of opening a test branch and enabling more logging to pinpoint the culprit.

**Comment by mokagio** [comment]
> My gut feeling is that the app is not being launched, so we'd need to perform the debugging directly there.

Looking at the trace uploaded in CI, it seems like the app launched alright:

Screenshot 2024-06-04 at 12 27 10

That is, if one assumes the white window is an unrelated issue. For what is worth, I see the white window locally when using `npm run e2e -- --ui`.

**Comment by mokagio** [comment]
The CI logs also show what should be an easy to address issue, which may be masking a different unit test failure:

```
[2024-06-03T12:46:53Z]     ReferenceError: fail is not defined
  | [2024-06-03T12:46:53Z]
  | [2024-06-03T12:46:53Z]       69 \| 					await fs.rm( dir, { recursive: true } );
  | [2024-06-03T12:46:53Z]       70 \| 				} catch {
  | [2024-06-03T12:46:53Z]     > 71 \| 					fail( `Failed to clean up ${ dir }` );
  | [2024-06-03T12:46:53Z]          \| 					^
  | [2024-06-03T12:46:53Z]       72 \| 				}
```

**Comment by mokagio** [comment]
@fluiddot @wojtekn I might have found the issue...

I added `DEBUG=pw:browser` as suggested in the PlayWright docs https://playwright.dev/docs/ci#debugging-browser-launches and in the logs we find

> Error occurred in handler for 'createSite': Error: Site server started with Playground's 'index' mode. Studio only supports 'wordpress' mode

Full logs:


```
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err] Error occurred in handler for 'createSite': Error: Site server started with Playground's 'index' mode. Studio only supports 'wordpress' mode. +2s
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err]     at S.<anonymous> (C:\buildkite-agent\builds\ci-windows-i-0a56dde48667a6ccb-1\automattic\studio\out\Studio-win32-x64\resources\app.asar\.webpack\main\index.js:2:2038597) +0ms
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err]     at Generator.next (<anonymous>) +0ms
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err]     at o (C:\buildkite-agent\builds\ci-windows-i-0a56dde48667a6ccb-1\automattic\studio\out\Studio-win32-x64\resources\app.asar\.webpack\main\index.js:2:2036301) +0ms
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err]     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) +0ms
[2024-06-04T03:36:32Z]   pw:browser [pid=4936][err] Sentry Logger [error]: Transport disabled +11ms
[2024-06-04T03:36:57Z] T°°  pw:browser [pid=4936] <gracefully close start> +26s
```

**Comment by mokagio** [comment]
The error above got me curious about what other differences in testing there might be between Windows and Linux or macOS.

I added a step to run the unit tests on Windows in CI and it's reporting [a few failures](https://buildkite.com/automattic/studio/builds/888#018fe15f-d7de-45de-bb36-6fba7c111b03).

**Comment by fluiddot** [comment]
> [...] My gut feeling is that the app is not being launched, so we'd need to perform the debugging directly there. I'm thinking of opening a test branch and enabling more logging to pinpoint the culprit.

Yes, I was wrong about the app being launched 😅. I also thought of the potential problem related to the timeout set for site creation. On Windows, due to Windows Defender, it takes way more time to create/run a site. Hence, it can easily exceed the 30-seconds threshold set [here](https://github.com/Automattic/studio/blob/8fc3a5dd024a194f016a0a8e286be19de84c32af/e2e/sites.test.ts#L50).

> I added `DEBUG=pw:browser` as suggested in the PlayWright docs https://playwright.dev/docs/ci#debugging-browser-launches and in the logs we find
> 
> > Error occurred in handler for 'createSite': Error: Site server started with Playground's 'index' mode. Studio only supports 'wordpress' mode

Great, good catch @mokagio! Based on the error, seems there's something off in the site folder. However, this seem to only happen on CI as I can't reproduce it on a Windows machine 🤔.

> I added a step to run the unit tests on Windows in CI and it's reporting [a few failures](https://buildkite.com/automattic/studio/builds/888#018fe15f-d7de-45de-bb36-6fba7c111b03).

That's interesting. I've run the unit tests on a Windows machine and encountered the same errors. We'd probably need to mock some parts of the app to support Windows. I can take a look 👀 .

**Comment by wojtekn** [comment]
> Great, good catch @mokagio! Based on the error, seems there's something off in the site folder. However, this seem to only happen on CI as I can't reproduce it on a Windows machine 🤔.

@fluiddot may it be an issue in the `inferMode` where `isWordPressDevelopDirectory` uses unix path separators?

**Comment by fluiddot** [comment]
> @fluiddot may it be an issue in the `inferMode` where `isWordPressDevelopDirectory` uses unix path separators?

This is not happening when testing locally on a Windows machine, so not sure if it would be related to the path calculation. Although this could behave differently in CI instance.

**Comment by fluiddot** [comment]
I've added some logs in [a testing branch](https://github.com/Automattic/studio/compare/test/windows-e2e-tests) I created to debug these issues, and noticed that the site folder has unexpected files:

```
Project path: C:\Users\BUILDK~1\AppData\Local\Temp\My-WordPress-Website
Site files: [
  'index.php',
  'license.txt',
  'readme.html',
  'wp-activate.php',
  'wp-admin'
]
```

Since the project path is not unique in the temporal folder, I wonder if a previous run in the same instance could be the culprit. I'm going to check further and potentially insert a unique ID to the E2E tests when creating the site.

**Comment by fluiddot** [comment]
Looks like the problem comes from the server files, I logged its content and it doesn't look like a full WordPress site ([reference](https://buildkite.com/automattic/studio/builds/905#018fe41e-3267-400d-8fe0-cac552884406/2129-2364)) 🤔. I'm not sure how it ended up to be with that content. I'm going to test cleaning up that folder for E2E tests as a workaround.

**UPDATE:** It fails even after cleaning up the server files folder. Seems the bundled copy has the wrong files. I'll investigate if there's any issue during the download after installing the dependencies.

**UPDATE 2:** The downloaded version has all the expected files ([reference](https://buildkite.com/automattic/studio/builds/907#018fe452-b016-4924-9891-a860588f29eb/236-2087)). Hence, there seems to be an issue either when packaging the app or copying the bundled WP version.

**Comment by mokagio** [comment]
Looks like there's more Windows-specific investigation to do here.

@wojtekn @fluiddot, if that's okay with you, I'd like to merge the macOS work in the meantime. See #207.

**Comment by wojtekn** [comment]
Thanks @mokagio. I merged the other branch, merged trunk to this one, and changed PR title to account for its reduced scope.

**Comment by fluiddot** [comment]
> @wojtekn @fluiddot, if that's okay with you, I'd like to merge the macOS work in the meantime. See #207.

Sounds good. Thanks @wojtekn for reviewing and merging it 🙇 .

**Comment by fluiddot** [comment]
I managed to reproduce the issue by downloading the binary generated in CI. However, it only fails when running E2E tests. If I open the executable and create a site, it copies all WordPress files 🙃. 

After adding more logs, I realized that the problem is related to not waiting for the server files to be copied before using the app. In general, this is not a problem as by the time you interact with the app the copy operation is finished. [This log entry](https://buildkite.com/automattic/studio/builds/934#018fe7a5-9731-4de3-89d2-b138f72770e8/1962-2240) ([reference](https://github.com/Automattic/studio/blob/98b90f5efb769b8b9a5718384c613cf00a22db75/src/setup-wp-server-files.ts#L39-L49)) should have happened before creating the site.

In e2e tests actions are executed as soon as the window is ready, so the combination of copy operation being a bit slower on Windows and creating a site right after the app is opened is resulting in a race condition problem. I'll solve this by moving the `setupWPServerFiles` call to be invoked when the app is ready but before the main window is created. This way we'll ensure that all server files are copied.

**Comment by fluiddot** [comment]
I managed to address the issue about the site folder but the test case `delete site` is still failing. As far as I've checked, there seems to be an error on that specific action. I'll keep investigating...

**UPDATE:** As expected, the delete action is failing with the error `Error: Failed to perform delete operation`.

**Comment by mokagio** [comment]
Great job @fluiddot !

> I managed to address the issue about the site folder but the test case delete site is still failing. As far as I've checked, there seems to be an error on that specific action. I'll keep investigating...

At least it's progress 😄

**Comment by mokagio** [comment]
@fluiddot 

> I realized that the problem is related to not waiting for the server files to be copied before using the app. In general, this is not a problem as by the time you interact with the app the copy operation is finished. [This log entry](https://buildkite.com/automattic/studio/builds/934#018fe7a5-9731-4de3-89d2-b138f72770e8/1962-2240) ([reference](https://github.com/Automattic/studio/blob/98b90f5efb769b8b9a5718384c613cf00a22db75/src/setup-wp-server-files.ts#L39-L49)) should have happened before creating the site.
>
> In e2e tests actions are executed as soon as the window is ready, so the combination of copy operation being a bit slower on Windows and creating a site right after the app is opened is resulting in a race condition problem. I'll solve this by moving the setupWPServerFiles call to be invoked when the app is ready but before the main window is created. This way we'll ensure that all server files are copied.

Just for my understanding, the fix you implemented is at the app runtime level, not in the E2E tests, right?

If I understand correctly, the unusual setup in the tests brought to light a design flaw that was present but hard to run into when interacting with the app as a human. If that's the case, I'll add this to my collection of examples of how testing helps discover edge cases 😄 

(If instead I have misunderstood, then that's good too as it's an occasion for me to learn more)

**Comment by fluiddot** [comment]
> Just for my understanding, the fix you implemented is at the app runtime level, not in the E2E tests, right?
>
> If I understand correctly, the unusual setup in the tests brought to light a design flaw that was present but hard to run into when interacting with the app as a human. If that's the case, I'll add this to my collection of examples of how testing helps discover edge cases 😄
> 
> (If instead I have misunderstood, then that's good too as it's an occasion for me to learn more)

Exactly, [here's the fix](https://github.com/Automattic/studio/commit/98b90f5efb769b8b9a5718384c613cf00a22db75) I applied. The issue is more prone to happen on E2E due to the fact that UI interaction happens quickly. Still, I presume it could be reproducible if you open the app and click on the "Add Site" button right away. Additionally, if disk operations are slow in the machine, this would be also easier to experience.

**Comment by fluiddot** [comment]
I tried different approaches to solve the error when deleting a site but none have proven to work ([reference](https://buildkite.com/automattic/studio/builds?branch=test%2Fwindows-e2e-tests)). I tested locally and all E2E tests pass, so I'm thinking of disabling that test case temporarily on Windows. WDYT?

**Comment by wojtekn** [comment]
> I tested locally and all E2E tests pass, so I'm thinking of disabling that test case temporarily on Windows. WDYT?

@fluiddot makes sense, let's do it. We can have a proper comment there and come back to that in the future, and still benefit from running other tests on Windows.

**Comment by fluiddot** [comment]
I continued exploring the needed workarounds to unblock this PR and managed the E2E tests to pass in a separate branch with the following changes:
- [Fix race condition when setting up server files](https://github.com/Automattic/studio/pull/229#top): As mentioned [here](https://github.com/Automattic/studio/pull/76#issuecomment-2149343524), the fact that you can create a site before the operation of copying the WordPress server files finishes might result in an exception. This should be uncommon as the copy operation should be fast. However, I noticed in the Windows CI machine that copying the bundled copy of WordPress can take around 30 seconds. Probably, this is potentially related to Microsoft Defender slowing down the process.
- [Increase E2E timeouts](https://github.com/Automattic/studio/pull/76/commits/2382be912b81279457c75142c739c01e33ca0518): Some of the actions like creating a site take longer on Windows, hence we need to increase the timeouts to avoid the failure.
- [Skip delete site test on Windows](https://github.com/Automattic/studio/pull/76/commits/09818dc8dbe7e48cedab98013ba67dc48932f0e8): As shared [here](https://github.com/Automattic/studio/pull/76#issuecomment-2152977538), I couldn't find a solution for the delete site test case. Until we find a workaround, we'll temporarily disable it.

Since this PR is related to E2E, I'm also going to apply the following changes:
- Re-enable E2E tests on Windows (seems they were disabled after updating with `trunk`).
- Disable unit tests temporarily on Windows until we find a solution to the failures.

cc @mokagio

**Comment by fluiddot** [comment]
@mokagio I took the liberty to set the PR ready for review as all CI jobs are passing (https://github.com/Automattic/studio/pull/76/commits/816fbb95740270e5868e213aba034705ac487b06) 🎊 . Let me know if you have any thoughts/concerns about this, thanks 🙇 !

**Review by fluiddot — APPROVED** [review]

**Comment by mokagio** [comment]
Excellent! Thanks @fluiddot
